File: cpp/sieve/sieve.cpp
#include "sieve.h"

namespace sieve {

auto primes(int limit) -> std::vector<int>
{
    std::vector<int> result;
    
    if (limit >= 2) {
        std::vector<bool> marks;
        marks.resize(limit - 1);
        
        for (int i = 2; i <= limit; ++i) {
            if (!marks[i - 2]) {
                result.push_back(i);
                for (int j = i + i; j <= limit; j += i) {
                    marks[j - 2] = true;
                }
            }
        }
    }
    
    return result;
}

}  // namespace sieve

File: r/sieve/sieve.R
sieve <- function(limit) {
  if (limit >= 2) {
    primes <- 2:limit
    n <- 2
    while (n < limit) {
      primes <- primes[!primes %in% seq(from = n * 2, to = limit * 2, by = n)]
      repeat {
        n <- n + 1
        if (n >= limit | n %in% primes) {
          break()
        }
      }
    }
    primes
  } else {
    c()
  }
}
